### PR TITLE
Paginate reactions - solved #1007

### PR DESCRIPTION
**Pull request by HelpfulStranger999**
# Changes
This pull requests changes the nature of IUserMessage#GetReactionUsersAsync to paginate the discord endpoint, thereby solving #1007. However, the signature of this method has been changed according, meaning all calls will now be broken.

Ignore the other commits where my inexperience with Git was at work.

### CODE REVIEW

File: src/Discord.Net.Core/Entities/Messages/EmbedBuilder.cs [hunk 1]
```diff
@@ -231,7 +231,6 @@ public class EmbedFieldBuilder
     {
         private string _name;
         private string _value;
-        private EmbedField _field;
```
Comment (SubZero0) [inline src/Discord.Net.Core/Entities/Messages/EmbedBuilder.cs]: Unrelated.

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 2]
```diff
@@ -46,13 +46,39 @@ internal static class MessageHelper
...
             Action<GetReactionUsersParams> func, BaseDiscordClient client, RequestOptions options)
         {
             var args = new GetReactionUsersParams();
             func(args);
-            string emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
-            return (await client.ApiClient.GetReactionUsersAsync(msg.Channel.Id, msg.Id, emoji, args, options).ConfigureAwait(false)).Select(u => RestUser.Create(client, u)).ToImmutableArray();
+
+            var emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
+            return new PagedAsyncEnumerable<IUser>(
+                DiscordConfig.MaxUsersPerBatch,
```
Comment (SubZero0) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: This value is wrong, it's 1000.
Create a new const (example: MaxUserReactionsPerBatch) with a value of 100 (don't reuse others or they might change and break this).

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 3]
```diff
@@ -46,13 +46,39 @@ internal static class MessageHelper
...
+                async (info, ct) =>
+                {
+                    if (info.Position != null)
+                        args.AfterUserId = info.Position.Value;
```
Comment (SubZero0) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: The correct way imo would be creating a local "args" with the page size and the position.

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 4]
```diff
@@ -46,13 +46,39 @@ internal static class MessageHelper
...
+                        builder.Add(RestUser.Create(client, model));
+
+                    return builder.ToImmutable();
+                },
+                nextPage: (info, lastPage) =>
+                {
+                    if (lastPage.Count != DiscordConfig.MaxUsersPerBatch)
+                        return false;
+
+                    info.Position = lastPage.OrderBy(u => u.Id).First().Id;
```
Comment (SubZero0) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: Use `Max(x => x.Id)`.

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 5]
```diff
@@ -46,13 +46,39 @@ internal static class MessageHelper
...
+                    return true;
+                },
+                count: args.Limit.Value
```
Comment (SubZero0) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: Shouldn't there be a `start:` here with the AfterUserId value?

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 6]
```diff
@@ -46,13 +46,39 @@ internal static class MessageHelper
...
-        public static async Task<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IMessage msg, IEmote emote,
+        public static IAsyncEnumerable<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IMessage msg, IEmote emote,
             Action<GetReactionUsersParams> func, BaseDiscordClient client, RequestOptions options)
         {
             var args = new GetReactionUsersParams();
             func(args);
-            string emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
-            return (await client.ApiClient.GetReactionUsersAsync(msg.Channel.Id, msg.Id, emoji, args, options).ConfigureAwait(false)).Select(u => RestUser.Create(client, u)).ToImmutableArray();
+
+            var emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
```
Comment (SubZero0) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: Better check for nulls before (msg and emote).
This will throw if emote is null.

**Review by SubZero0 — COMMENTED** [review]
You should also change `DiscordRestApiClient.GetReactionUsersAsync`.

- Change the `MaxUsersPerBatch` there to your new const (that I said in the review).
- Change the `int.MaxValue` to your new const as well.
- Add `?limit={limit}&after={afterUserId}` to the endpoint string (or reference the other PR that only does this, this PR alone won't work).

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs (hunk 6 above)
Comment (HelpfulStranger999) [reply src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: The message shouldn't be null, since the whole class is internal, and the places where it is called (Message objects) can't be null else Nullrefs will be thrown before it enters that method. So is it strictly necessary?

Comment (HelpfulStranger999) [reply src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: Or is it standard throughout the rest of the library?

Comment (SubZero0) [reply src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: You do have a point about the message, but `emote` could still receive null and throw.

File: src/Discord.Net.Rest/DiscordRestApiClient.cs [hunk 7]
```diff
@@ -618,15 +618,15 @@ internal virtual void Dispose(bool disposing)
...
             Preconditions.GreaterThan(args.Limit, 0, nameof(args.Limit));
-            Preconditions.AtMost(args.Limit, DiscordConfig.MaxUsersPerBatch, nameof(args.Limit));
+            Preconditions.AtMost(args.Limit, DiscordConfig.MaxUserReactionsPerBatch, nameof(args.Limit));
             Preconditions.GreaterThan(args.AfterUserId, 0, nameof(args.AfterUserId));
             options = RequestOptions.CreateOrClone(options);
 
-            int limit = args.Limit.GetValueOrDefault(int.MaxValue);
-            ulong afterUserId = args.AfterUserId.GetValueOrDefault(0);
+            var limit = args.Limit.GetValueOrDefault(DiscordConfig.MaxUserReactionsPerBatch);
+            var afterUserId = args.AfterUserId.GetValueOrDefault(0);
```
Comment (SubZero0) [inline src/Discord.Net.Rest/DiscordRestApiClient.cs]: I'm pretty sure the standard for this lib is to only use `var` when the type is explicit.

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 8]
```diff
@@ -46,13 +46,45 @@ internal static class MessageHelper
...
+            var arguments = new GetReactionUsersParams();
+            func(arguments);
+
+            return new PagedAsyncEnumerable<IUser>(
+                DiscordConfig.MaxUserReactionsPerBatch,
+                async (info, ct) =>
+                {
+                    var args = new GetReactionUsersParams();
+                    func(args);
+
```
Comment (SubZero0) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: `args.Limit = info.PageSize;`
To request the correct amount.

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 9]
```diff
@@ -46,13 +46,45 @@ internal static class MessageHelper
...
+                },
+                nextPage: (info, lastPage) =>
+                {
+                    if (lastPage.Count != DiscordConfig.MaxUsersPerBatch)
+                        return false;
+
+                    info.Position = lastPage.Max(x => x.Id);
+                    return true;
+                },
+                start: arguments.AfterUserId.Value,
```
Comment (SubZero0) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: This can throw.
`start: arguments.AfterUserId.IsSpecified ? args.AfterUserId.Value : (ulong?)null,`

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 10]
```diff
@@ -46,13 +46,45 @@ internal static class MessageHelper
...
+                count: arguments.Limit.Value
```
Comment (SubZero0) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: This could also throw but it's harder since it's internal. The best way would be:
`count: args.Limit.GetValueOrDefault(DiscordConfig.MaxUserReactionsPerBatch)`

File: src/Discord.Net.Rest/DiscordRestApiClient.cs (hunk 7 above)
Comment (HelpfulStranger999) [reply src/Discord.Net.Rest/DiscordRestApiClient.cs]: Ohhh, I see, ok

File: src/Discord.Net.Core/Entities/Messages/IUserMessage.cs [hunk 11]
```diff
@@ -23,7 +23,7 @@ public interface IUserMessage : IMessage
         /// <summary> Removes all reactions from this message. </summary>
         Task RemoveAllReactionsAsync(RequestOptions options = null);
         /// <summary>  Gets all users that reacted to a message with a given emote </summary>
-        Task<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IEmote emoji, int limit = 100, ulong? afterUserId = null, RequestOptions options = null);
+        IAsyncEnumerable<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IEmote emoji, int limit = 100, ulong? afterUserId = null, RequestOptions options = null);
```
Comment (FiniteReality) [inline src/Discord.Net.Core/Entities/Messages/IUserMessage.cs]: Make limit required and remove afterUserId

File: src/Discord.Net.WebSocket/Entities/Messages/SocketUserMessage.cs [hunk 12]
```diff
@@ -130,7 +130,7 @@ public Task RemoveReactionAsync(IEmote emote, IUser user, RequestOptions options
             => MessageHelper.RemoveReactionAsync(this, user, emote, Discord, options);
         public Task RemoveAllReactionsAsync(RequestOptions options = null)
             => MessageHelper.RemoveAllReactionsAsync(this, Discord, options);
-        public Task<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IEmote emote, int limit = 100, ulong? afterUserId = null, RequestOptions options = null)
+        public IAsyncEnumerable<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IEmote emote, int limit = 100, ulong? afterUserId = null, RequestOptions options = null)
```
Comment (FiniteReality) [inline src/Discord.Net.WebSocket/Entities/Messages/SocketUserMessage.cs]: Make limit required and remove afterUserId

File: src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs [hunk 13]
```diff
@@ -136,7 +136,7 @@ public Task RemoveReactionAsync(IEmote emote, IUser user, RequestOptions options
             => MessageHelper.RemoveReactionAsync(this, user, emote, Discord, options);
         public Task RemoveAllReactionsAsync(RequestOptions options = null)
             => MessageHelper.RemoveAllReactionsAsync(this, Discord, options);
-        public Task<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IEmote emote, int limit = 100, ulong? afterUserId = null, RequestOptions options = null)
+        public IAsyncEnumerable<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IEmote emote, int limit = 100, ulong? afterUserId = null, RequestOptions options = null)
```
Comment (FiniteReality) [inline src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs]: Make limit required and remove afterUserId

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 14]
```diff
@@ -46,13 +46,46 @@ internal static class MessageHelper
             await client.ApiClient.RemoveAllReactionsAsync(msg.Channel.Id, msg.Id, options);
         }
 
-        public static async Task<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IMessage msg, IEmote emote,
+        public static IAsyncEnumerable<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IMessage msg, IEmote emote,
```
Comment (FiniteReality) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: Remove GetReactionUserParams func here and add `ulong? from`, `int? limit` params instead

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 15]
```diff
@@ -46,13 +46,46 @@ internal static class MessageHelper
...
+            var emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
+
+            var arguments = new GetReactionUsersParams();
+            func(arguments);
+
+            return new PagedAsyncEnumerable<IUser>(
+                DiscordConfig.MaxUserReactionsPerBatch,
+                async (info, ct) =>
+                {
+                    var args = new GetReactionUsersParams();
```
Comment (FiniteReality) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: Replace with ctor which sets Limits here, and remove func invocation

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 16]
```diff
@@ -46,13 +46,46 @@ internal static class MessageHelper
...
+                    func(args);
+                    args.Limit = info.PageSize;
+
+                    if (info.Position != null)
+                        args.AfterUserId = info.Position.Value;
+
+                    var models = await client.ApiClient.GetReactionUsersAsync(msg.Channel.Id, msg.Id, emoji, args, options).ConfigureAwait(false);
+                    var builder = ImmutableArray.CreateBuilder<IUser>();
```
Comment (FiniteReality) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: IMO this should be replaced with models.Select and you should call ToImmutableArray on it

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 17]
```diff
@@ -46,13 +46,46 @@ internal static class MessageHelper
...
             Action<GetReactionUsersParams> func, BaseDiscordClient client, RequestOptions options)
         {
-            var args = new GetReactionUsersParams();
-            func(args);
-            string emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
-            return (await client.ApiClient.GetReactionUsersAsync(msg.Channel.Id, msg.Id, emoji, args, options).ConfigureAwait(false)).Select(u => RestUser.Create(client, u)).ToImmutableArray();
+            Preconditions.NotNull(emote, nameof(emote));
+            var emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
+
+            var arguments = new GetReactionUsersParams();
```
Comment (FiniteReality) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: Remove this and the func invocation

File: src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs [hunk 18]
```diff
@@ -46,13 +46,46 @@ internal static class MessageHelper
...
+                nextPage: (info, lastPage) =>
+                {
+                    if (lastPage.Count != DiscordConfig.MaxUsersPerBatch)
+                        return false;
+
+                    info.Position = lastPage.Max(x => x.Id);
+                    return true;
+                },
+                start: arguments.AfterUserId.IsSpecified ? arguments.AfterUserId.Value : (ulong?)null,
+                count: arguments.Limit.GetValueOrDefault(DiscordConfig.MaxUserReactionsPerBatch)
```
Comment (FiniteReality) [inline src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs]: Replace these with from and limit suggested earlier

**Review by FiniteReality — CHANGES_REQUESTED** [review]

File: src/Discord.Net.Core/Entities/Messages/IUserMessage.cs (hunk 11 above)
Comment (SubZero0) [reply src/Discord.Net.Core/Entities/Messages/IUserMessage.cs]: Why make it required when we dont do that for `GetMessagesAsync` that follow a similar pattern? (Could change the `= 100` to `= DiscordConfig.MaxUserReactionsPerBatch` tho)
And why remove an option that the discord api gives to us?

Comment (FiniteReality) [reply src/Discord.Net.Core/Entities/Messages/IUserMessage.cs]: It looks like GetMessagesAsync is the only one where we have a default: everywhere else limit is required (and even in IMessageChannel, limit is required!) or is not present at all.

File: src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs (hunk 13 above)
Comment (HelpfulStranger999) [reply src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs]: Which should afterUserId go then?

Comment (FiniteReality) [reply src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs]: It's handled as part of the async enumerable returned, so it's unnecessary.

Comment (HelpfulStranger999) [reply src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs]: How does the parameter get passed to MessageHelper.GetReactionUsersAsync then?

Comment (FiniteReality) [reply src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs]: It doesn't need to be; our async enumerable code handles it automatically.

Comment (HelpfulStranger999) [reply src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs]: Ohhhh, I see what you mean, where do I get the starting point for the Paged Enumerable then or should I leave that out?

Comment (FiniteReality) [reply src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs]: The paged enumerable defaults to 0, which would retrieve all of the results anyway.

**Comment by HelpfulStranger999** [comment]
@FiniteReality Should be ready for you to re-review

**Review by FiniteReality — APPROVED** [review]
